### PR TITLE
Contract storage slots checker

### PR DESCRIPTION
**Pull request by sparrowDom**
This PR checks before a contract is upgraded that it doesn't corrupt the storage slot layout of the currently live contract. To achieve this: 
 - When a contract gets deployed its storage slot layout is stored to `storageLayout/[network]/[contractName].json`. And when a user deploys contracts to mainnet those layout changes are reflected in repo and should be pushed to master. 
 - On a contract upgrade the currently live storage layout (stored in repo) and a new layout derived from contract source code are compared and if there is a conflict an exception is thrown.

**Connected issue:** https://github.com/OriginProtocol/origin-dollar/issues/468

**A few helpful hardhat tasks are added:**
 - saveStorageSlotLayout -> saves storage layout for all contracts
 - checkUpgradability -> checks upgradability of a specific contract
 - checkUpgradabilityAll -> checks upgradability for all contracts 

The current storage slot json files have been generated from contract code commit done on Jan 8th, when the 011_ousd_fix.js has been created. 

**Example of storage slot corruption detection:**

Foo variable is added to OUSD contract
Screenshot 2021-01-29 at 21 40 10

Upon launching the upgrade an error is thrown: 
Screenshot 2021-01-29 at 21 39 58

For solidity gurus: 
 - is excluding [Proxy and the following contracts](https://github.com/OriginProtocol/origin-dollar/pull/536/files#diff-92f367937ac4637af4322bf6c4e30970fc0f6ffea938f08d1173c60a101571dcR77) from storage slot checker ok?
 
Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [ ] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

**Comment by sparrowDom** [comment]
Is excluding these contract names from slot check ok?
https://github.com/OriginProtocol/origin-dollar/pull/536/files#diff-92f367937ac4637af4322bf6c4e30970fc0f6ffea938f08d1173c60a101571dcR77

Also we could be using `hardhat-upgrades` module that does these storage slot checks for us, but the initial contract proxy needs to be created using the `hardhat-upgrades` code. There is an [open issue on their repo](https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/175), to allow registering a proxy even if it was not created using their code, but it has not been implemented yet.

File: contracts/tasks/storageSlots.js
Comment (franckc) [inline contracts/tasks/storageSlots.js:69]: I would add a comment explaining those contracts have been deprecated. They remain present in the deployment files but the contract code has been deleted and therefore they need to be excluded.

File: contracts/tasks/storageSlots.js
Comment (franckc) [inline contracts/tasks/storageSlots.js:77]: Proxies are not upgradeable so make sense to exclude them.

File: contracts/utils/deploy.js
Comment (franckc) [inline contracts/utils/deploy.js:41]: Rather than here. should this happen once the deploy is done and confirmed. Like after the call to withConfirmation below? So that in case of deploy error, the storage layout is not persisted.

File: contracts/tasks/storageSlots.js
Comment (franckc) [inline contracts/tasks/storageSlots.js:12]: Don't we need to add this package to package.json? We currently have @openzeppelin/upgrades in there but not @openzeppelin/upgrades-core

**Review by franckc — COMMENTED** [review]
I'm super excited we are adding this tool! Thank you for doing that.

A few minor comments inline.

File: contracts/tasks/storageSlots.js
Comment (sparrowDom) [reply contracts/tasks/storageSlots.js:69]: Aha that makes sense thanks.

File: contracts/tasks/storageSlots.js
Comment (sparrowDom) [reply contracts/tasks/storageSlots.js:77]: 👍

File: contracts/tasks/storageSlots.js
Comment (sparrowDom) [reply contracts/tasks/storageSlots.js:12]: I think it worked for me because another package already includes it. Thanks for catching this, added to `contracts/package.json`

File: contracts/utils/deploy.js
Comment (sparrowDom) [reply contracts/utils/deploy.js:41]: Yup makes a lot more sense, moving to there

**Comment by sparrowDom** [comment]
All comments addressed. Will stick to our soft defined deal and wait for 2 CR confirmations before merging this one in.

**Review by franckc — APPROVED** [review]
Thanks for addressing my comments. Nice work on this! :) LGTM

Lets merge this in and use it when running deploy 12

**Comment by franckc** [comment]
@sparrowDom Changes look good. Though can you take a look at those tests failures before merging this in?

**Comment by sparrowDom** [comment]
Disabling this functionality on tests for now. Reading hardhat cache fails in CI and not in local test environment for some reason. Do object if you would like to see the storage slot checker working on test env anyway.

**Pull request merged**
